### PR TITLE
Remove Sums Class

### PR DESCRIPTION
**Pull request by aniketd**
# Description

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [x] The version bounds in `.cabal` files for all affected packages are updated (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/input-output-hk/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Spec.hs [hunk 1]
```diff
@@ -1482,15 +1467,18 @@ mergeElemSpec a@(ElemEqual _ xs) b@(ElemSum _ sz) =
...
+-- mergeElemSpec a@(ElemProj sm1 r1 l1 sz1) b@(ElemProj sm2 r2 l2 sz2) =
+--   case testEql r1 r2 of
+--     Just Refl ->
+--       case sz1 <> sz2 of
+--         SzNever xs -> ElemNever (sepsP ["The ElemSpec's are inconsistent.", show a, show b] : xs)
+--         sz3 ->
+--           if l1 == l2
+--             then ElemProj (smallerOf sm1 sm2) r1 l1 sz3
+--             else ElemNever ["The lenses for ElemProj are inconsistent", " " <> show a, " " <> show b]
+--     Nothing -> ElemNever ["The ElemSpec's are inconsistent.", "  " ++ show a, "  " ++ show b]
```
Comment (aniketd) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Spec.hs]: @TimSheard There seems to be know way to merge the specs once we use lenses, am I correct?

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs [hunk 2]
```diff
@@ -88,46 +88,55 @@ import Test.QuickCheck (
...
   minus :: [String] -> x -> x -> x
-  zero :: x
 
-  -- | 'partition 7 trace 4 235', generate a list of length 4 that
-  --   adds up t0 235, where the smallest number is >= 7
+  -- | Increase by unit of increment
+  increase :: x -> x
+  default increase :: x -> x
+  increase = add one
+
```
Comment (TimSheard) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs]: increaseBy1  decreaseBy1

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs [hunk 3]
```diff
@@ -88,46 +88,55 @@ import Test.QuickCheck (
...
+  -- | Decrease by unit of increment
+  decrease :: [String] -> x -> x
+  default decrease :: [String] -> x -> x
+  decrease msgs = minus msgs one
+
+  -- | Negate a value
+  negateAdds :: x -> x
+
```
Comment (TimSheard) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs]: zero - x

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs [hunk 4]
```diff
@@ -137,61 +146,59 @@ class (Eq x, Show x) => Adds x where
...
+genFromNonNegAddsSpec msgs (AddsSpecNever _) = errorMess "genFromAddsSpec applied to AddsSpecNever" msgs
 
 instance Adds ExUnits where
+  zero = ExUnits 0 0
+  one = ExUnits 1 1
   add (ExUnits a b) (ExUnits c d) = ExUnits (a + c) (b + d)
   minus msgs (ExUnits a b) (ExUnits c d) =
     ExUnits
       (minus ("Ex memory" : msgs) a c)
       (minus ("Ex steps" : msgs) b d)
```
Comment (TimSheard) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:188]: minus on ExUnits memory

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs [hunk 5]
```diff
@@ -312,20 +324,15 @@ instance Adds DeltaCoin where
...
-  genT _ cn = genTxOutX reify cn
+-- instance Adds Size where
+--   zero = SzNever ["Adds Size (zero)"]
+--   one = SzRng 1 1
+--   add = mergeSize
+--   minus msgs _ _ = SzNever msgs
+--   negateAdds = negateSize
+--   genAdds msgs spec = fromI ms <$> genFromAddsSpec ms spec
+--     where
+--       ms = msgs ++ ["genAdds DeltaCoin"]
```
Comment (TimSheard) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs]: No need for this.

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs [hunk 6]
```diff
@@ -932,3 +923,159 @@ partitionWord64 small msgs n total =
...
+
+-- | Translate (s,cond,n), into a Size which
+--   specifies the Int range on which the OrdCond is True.
+--   The triple (s, EQL, 2) denotes s = 2
+--              (s, LTH, 7) denotes s < 7
+--              (s, GTH, 5) denotes s > 5 ...
+ordCondToSize :: Adds a => (String, OrdCond, a) -> Size
+ordCondToSize (label, cond, n) = case cond of
+  EQL -> SzExact $ toI n
+  LTH -> SzMost $ toI $ decrease [unlines ["ordCondToSize failed for", label, show n]] n
```
Comment (TimSheard) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs]: decrease n = pair (i-1) (j-1)  where (i,j) = unpair n


EQL -> SzExact $ toI n
LTH -> SzMost $ decrease [unlines ["ordCondToSize failed for", label, show n]] (toI n)
 GTH -> SzLeast $ increase (toI n)

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs [hunk 7]
```diff
@@ -932,3 +923,159 @@ partitionWord64 small msgs n total =
...
+-- genAdds :: Adds c => Size -> Gen x
+-- genAdds (SzExact n)
+--   where (i, j) = unpair n, do ig <- genAdds (SzExact i), jg <- genAdds (szExact j), pure (ExUnits ig jg)
+
+-- Translate some thing like [SumsTo _ x <= 4 + 6 + 9] where the variable 'x' is on the left
+vLeft :: Adds a => String -> OrdCond -> a -> AddsSpec c
+vLeft x cond n = AddsSpecSize x (vLeftSize x cond n)
+
+vLeftSize :: Adds a => String -> OrdCond -> a -> Size
+vLeftSize x cond n = ordCondToSize (x, cond, n)
```
Comment (TimSheard) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs]: vLeft --> varOnLeft

vRightSize  --> varOnRightSize

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Spec.hs (hunk 1 above)
Comment (TimSheard) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Spec.hs]: yes

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Solver.hs [hunk 8]
```diff
@@ -27,24 +26,24 @@ import Numeric.Natural (Natural)
...
-  AddsSpec (..),
-  OrdCond (..),
   Size (..),
   genFromIntRange,
   genFromSize,
-  vLeft,
-  vRight,
-  vRightNeg,
-  vRightSize,
  )
```
Comment (TimSheard) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Solver.hs:49]: At all these call sites. One MUST get rid of the toI ,

**Review by TimSheard — CHANGES_REQUESTED** [review]

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs (hunk 3 above)
Comment (aniketd) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs]: Removed `negateAdds` altogether, following our conversation :+1:

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs (hunk 4 above)
Comment (aniketd) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Classes.hs:188]: I don't exactly remember what this comment was for, during our discussion @TimSheard. Can we discuss on Slack?

**Pull request merged**